### PR TITLE
Add normal distribution regression objective

### PR DESCRIPTION
**Pull request by RAMitchell**
## Summary

Adds `reg:normal`, a distributional regression objective that predicts two values per row:

- conditional mean
- log variance

The objective supports both one-tree-per-output and shared-topology vector trees. It also adds the `normal-nloglik` evaluation metric.

## Stable variance updates

For log variance $s$, define

$$
z = \left((y-\mu)^2 + \epsilon\right)e^{-s},
$$

where $\epsilon$ is float epsilon. The gradient and pseudo-Hessian are

$$
g_s = \frac{w}{2}(1-z),
\qquad
h_s = \frac{w}{6}(1+2z).
$$

The observed Hessian $wz/2$ vanishes when the residual approaches zero. This can produce an arbitrarily large Newton update for log variance.

The proposed curvature instead gives the fixed-mean leaf update

$$
d = \frac{3(R-1)}{1+2R},
$$

where $R$ is the leaf-average standardized squared residual. This update is bounded to $(-3, 3/2)$, moves toward the exact optimum $\log R$, and provides a conservative quadratic gain estimate.

Adding float epsilon gives the variance a finite optimum when the mean interpolates the observations. It prevents training from continually driving the variance toward zero without introducing a user parameter.

The vector intercept is initialized exactly as the weighted response mean and

$$
\log(\text{weighted residual variance}+\epsilon).
$$

## Results

The comparison used shared-topology vector trees on four public regression datasets, with three deterministic train/validation splits.

At `eta=0.3`, `lambda=1`, depth 4, and 100 rounds:

| Dataset | Newton validation NLL | Bounded-step validation NLL |
|---|---:|---:|
| Ailerons | 0.3648 | **0.3573** |
| CPU Activity | -0.6502 | **-0.7033** |
| Kin8nm | 0.8479 | **0.8218** |
| Wine Quality | 1.2002 | **1.1956** |

In a zero-regularization stress test at `eta=0.1`, all 12 observed-Newton runs became non-finite. All 12 bounded-step runs completed 100 rounds without a training-loss increase.

A separate interpolation stress test used `eta=1`, depth 12, and no regularization. All eight combinations of four synthetic datasets and the two multi-output strategies completed 100 rounds with finite predictions. Constant-label leaves settled at float epsilon variance.

## Implementation

- CPU and CUDA objective kernels
- exact vector-valued intercept
- `normal-nloglik` metric
- support for `multi_output_tree` and `one_output_per_tree`
- parameter documentation
- CPU, CUDA compilation, metric, and Python integration tests

### CODE REVIEW

File: src/metric/elementwise_metric.cu [hunk 1]
```diff
@@ -367,6 +367,43 @@ struct EvalEWiseBase : public MetricNoCache {
...
+    preds.SetDevice(device);
+    auto predts = ctx_->IsCUDA() ? preds.ConstDeviceSpan() : preds.ConstHostSpan();
+
+    constexpr float kLogTwoPi = 1.8378770664093453f;
+    auto result =
+        Reduce(ctx_, info, [=] XGBOOST_DEVICE(size_t, size_t sample_id, size_t target_id) {
+          auto mean = predts[sample_id * 2];
+          auto log_variance = predts[sample_id * 2 + 1];
+          auto residual = labels(sample_id, target_id) - mean;
+          auto loss = 0.5f * (kLogTwoPi + log_variance + residual * residual * expf(-log_variance));
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply src/metric/elementwise_metric.cu]: Fixed in 99b8c0a8f. The standardized residual is now evaluated in log space with an explicit zero-residual branch. The test covers both an exact zero residual and a small nonzero residual at log variance -100.

**Comment by trivialfis** [comment]
Are we going to support forecasting objectives?

**Comment by RAMitchell** [comment]
I had no plans for other objectives. This is useful outside of forecasting. If you have suggestions I can take a look.

**Review by trivialfis — APPROVED** [review]
